### PR TITLE
feat(compartment-mapper): Thread commonjs- and module- languageForExtension options (for TS)

### PR DESCRIPTION
**Pull request by kriskowal**
Refs: #2415

## Description

Toward support for type-erasure style TypeScript support in Endo, `.ts` files may correspond to `.mts` or `.cts` behavior in the same way that `.js` can correspond to either `.mjs` or `.cjs`, depending on `"type":` in the enveloping `package.json` (which is not necessarily the enveloping package, e.g., `src/package.json` beneath `package.json`!). We already implement this machinery for JavaScript, so this change threads additional options beside `languageForExtension` (which would map `.mts` and `.cts` to language behaviors) but also `commonjsLanguageForExtension` and `moduleLanguageForExtension`, which would get folded into the language-for-extension mapping on a package-by-package basis.

### Security Considerations

None. Absent from this design is any possibility that a file would be alternately implemented as TypeScript or JavaScript.

### Scaling Considerations

None.

### Documentation Considerations

Includes an update to README for the new features and the undocumented features it is based upon. Includes a mention in NEWS.

### Testing Considerations

TODO

### Compatibility Considerations

By design error, `languageForExtension` threaded through `link` before this change, when it should have been threaded through `mapNodeModules`, since that is the machine that is aware of the `type` field in `package.json`. This change moves that internal mechanism and then adds `commonjsLanguageForExtension` and `moduleLanguageForExtension`. By common usage, this move is transparent since the option is taken at the top and simply diverted internally.

However, the `languageForExtension` property could be manually threaded through `makeArchiveLite` and omitted from the options for `mapNodeModules`. I’ve elected to treat this breaking change as a bug fix since I find it unlikely anyone has taken advantage of these relatively new “lite” functions. The “lite” functions are part of the refactoring necessary to approach native XS compartment and modules.

### Upgrade Considerations

None.

### CODE REVIEW

File: packages/compartment-mapper/test/language-for-extension.test.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import jsonp from './_parse-jsonp.js';
+import text from '../src/parse-text.js';
+
+// The JSONP parser uses harden, as a bit.
+lockdown({
+  errorTaming: 'unsafe',
+  errorTrapping: 'none',
+});
+
+scaffold(
```
Comment (kriskowal) [inline packages/compartment-mapper/test/language-for-extension.test.js:13]: For reviewers new to Compartment Mapper, `scaffold` generates a combination of test cases where a single fixture is expected to produce the same behavior for multiple flows:

- executed directly off a file system
- loaded off the file system then later executed
- archived off the file system, then parsed, then executed
&c

File: packages/compartment-mapper/src/node-modules.js [hunk 2]
```diff
@@ -189,83 +196,73 @@ const findPackage = async (readDescriptor, canonical, directory, name) => {
...
-  'cjs',
-  'json',
-  'text',
-  'bytes',
-]);
-const defaultUncontroversialParsers = /** @type {const} */ ({
-  cjs: 'cjs',
+const defaultLanguageForExtension = /** @type {const} */ ({
   mjs: 'mjs',
+  cjs: 'cjs',
```
Comment (turadg) [inline packages/compartment-mapper/src/node-modules.js:203]: consider alphabetizing or leaving where it was

File: packages/compartment-mapper/src/node-modules.js [hunk 3]
```diff
@@ -189,83 +198,83 @@ const findPackage = async (readDescriptor, canonical, directory, name) => {
...
+    packageLanguageForExtension === null
+  ) {
+    throw Error(
+      `Cannot interpret parser map ${JSON.stringify(
+        packageLanguageForExtension,
+      )} of package at ${location}, must be an object mapping file extensions to corresponding languages (for example, mjs for ECMAScript modules, cjs for CommonJS modules, or json for JSON modules`,
+    );
+  }
+  const invalidLanguages = values(packageLanguageForExtension).filter(
+    language => !languages.has(language),
```
Comment (turadg) [inline packages/compartment-mapper/src/node-modules.js:256]: `language` and `languages` look very similar. consider,
```suggestion
    l => !languages.has(l),
```

**Review by turadg — APPROVED** [review]
LGTM!

I'm new to this package so I don't have an opinion on the design decisions. The code changes all look good.

**Pull request merged**